### ISSUE TITLE
Extra GPU-CPU memory transfer when broadcasting operations between integer tensors

### ISSUE THREAD

**Issue by unrealwill** [tensorflow/tensorflow#54197]
Hello,

<em>Please make sure that this is an issue related to performance of TensorFlow.
As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:performance_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: NA
- TensorFlow installed from (source or binary):binary
- TensorFlow version (use command below): v2.6.1-9-gc2363d6d025 2.6.2
- Python version: 3.6.9
- Bazel version (if compiling from source):NA
- GCC/Compiler version (if compiling from source):NA
- CUDA/cuDNN version: NVIDIA-SMI 495.29.05    Driver Version: 495.29.05    CUDA Version: 11.5
- GPU model and memory: GeForce 1080Ti

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

The code runs fine but when we look in detail in Tensorboard we observe that there are silent memory transfers between CPU and GPU, when doing operation that involve integer tensors that need broadcasting.

In the picture below (The green block "MemCpyH2D" in the Stream23, that has been clicked to show the name of the culprit op, should not exist) : 

![2022-01-29-175200_1920x1080_scrot](https://user-images.githubusercontent.com/11304248/151670751-2498d7a0-a26a-415e-9f2d-0c6f92312d12.png)

**Describe the expected behavior**
I expect 0 memory transfer for ops as simple as additions, bitwise element operation like bitwise_and with a mask that is a constant, integer reduction along various axis, expand_dims, reshape and squeeze, gather_nd and scatter_nd should also not result in data transfer between CPU and GPU

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

Create a log folder if necessary
And install tensorboard profiler (See https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras ) : 
`pip install -U tensorboard_plugin_profile`

```
import tensorflow as tf
from datetime import datetime

class CustomLayer(tf.keras.layers.Layer):
    def __init__(self):
        #N must be divisible by 4
        super(CustomLayer, self).__init__()
        self.dense1 = tf.keras.layers.Dense(100,use_bias=False)

    @tf.function
    def call(self, query):
        c = tf.cast(query,tf.int32,name="castQueryToInt32")
        d = tf.expand_dims( c, axis=1,name="expandDimsAxis1")
        e = tf.expand_dims( c, axis=2,name="expandDimsAxis2")
        g = tf.add(d , e ,name="additionBroadcasted")
        #h = tf.reduce_sum(g, axis=1) #This also create extra memory transfer between gpu and cpu
        f = tf.cast(g,tf.float32, name="castGToFloat")
        f = tf.reduce_sum(f,axis=1,name="reduceSumF")
        rem = (query - f)* (query - f)
        out = self.dense1(rem)
        return out



inputs = tf.keras.Input(shape=(100,))
out = CustomLayer()(inputs)


model = tf.keras.Model(inputs=inputs, outputs=out)
model.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=1e-3),loss="mse")


xtrain = tf.random.uniform((600000,100),dtype=tf.float32)
ytrain = tf.random.uniform((600000,100),dtype=tf.float32)

logs = "logs/" + datetime.now().strftime("%Y%m%d-%H%M%S")

#We investigate the slowness due the extra memory transfer between GPU and CPU because of broadcasting behavior of integer tensors
tboard_callback = tf.keras.callbacks.TensorBoard(log_dir=logs,
                                                 histogram_freq=1,
                                                 profile_batch="10,20")

model.fit( xtrain,ytrain, batch_size=100,epochs=1,callbacks=[tboard_callback])
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

I had simple code (additions and reductions between integer tensors of broadcastable shapes)  that had very low GPU utilization (25% and abysmal performance 0.1x), so I investigate with Tensorboard, observed plenty of non-necessary GPU-CPU transfer and extracted this simple example. (Once you fix this example, I can provide additional test cases).

Maybe some additional ops need to be registered/created to handle the integer tensors properly.

The ugly work-around is to cast every int32 to float64 do the operations that need broadcasting there and cast back to int32, it almost works (except for the backward pass of a gather_nd that I haven't managed to suppress yet).
I have tried using repeat with integer tensor but it result in extra transfer too.

Thank you

**Comment by unrealwill** [comment]
Additional related bug : 
When I try to multiply integer tensor of the same shape on the GPU (or by an integer scalar), they also get moved to the CPU.

I looked at the code and it seems the culprit is here : 
https://github.com/tensorflow/tensorflow/blob/fd16fa9fc6741bc7bb7fcc9dd70c1e527834c133/tensorflow/core/kernels/cwise_op_mul_1.cc#L39-L48

Can you also fix it please ?

Thanks

**Comment by sushreebarsa** [comment]
@unrealwill 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thanks!

**Comment by unrealwill** [comment]
@sushreebarsa Here is the cross-post : https://github.com/keras-team/keras/issues/15986 

Imho, this issue is not a Keras one but a tensorflow one, as it has nothing to do with the CustomLayer, but everything to do with the low level op like additions and multiplications not properly registered and defined for the GPU.

Thanks

**Comment by sushreebarsa** [comment]
@unrealwill Sorry for the late response!
As we see that you have raised this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues), please move this issue to closed status as we will track the other ticket there.Thanks!

**Comment by unrealwill** [comment]
@sushreebarsa No sorry, I won't close this thread here until it has been resolved as all the information is in this thread.
Furthermore, no one has answered the case there in 14 days, and the bug is not a Keras issue but a Tensorflow one.

Sending bugs to die in other hidden threads is not the way to increase your "closed case" performance metric. 

This bug is a basic low-level bug about integer tensors (we are talking about addition and multiplication here) that silently hit, with a huge performance impact, a lot of advanced users that try to use Tensorflow for state of the art research in everything that deals with sparsity and hashes. 

You should be glad I saw the hard to spot bug and reported it. You should have fixed it already !

**Comment by sushreebarsa** [comment]
@chunduriv Was able to reproduce on colab using TF v2.6.0 , 2.8.0 ,please find the attached gists for [cpu ](https://colab.research.google.com/gist/sushreebarsa/e3c1e12aabe6f9b551c0f6b40c997a24/reproduce.ipynb#scrollTo=ddQbVvIJAXTz)and [gpu](https://colab.research.google.com/gist/sushreebarsa/c56fb812c7b5697267425c2174c806fe/reproduce.ipynb#scrollTo=ddQbVvIJAXTz) .Thanks!

**Comment by rchao** [comment]
Wondering if you happened to try with `tf.device` to attempt to ping the ops on CPUs only? If not, that may be a reasonable solution to try.

**Comment by chunduriv** [comment]
@unrealwill,
Can you please check above [comment](https://github.com/tensorflow/tensorflow/issues/54197#issuecomment-1043283380) by @rchao and let us know? Thanks!

**Comment by unrealwill** [comment]
@chunduriv I tried to use tf.device("/GPU:0") to force the computation to happen on the GPU, (I don't care at all about the CPU performance) but it has no effect whatsoever.

**Comment by 372046933** [comment]
BTW, TF 2.9.1 `Reshape` OP also copies input tensor to CPU


https://github.com/tensorflow/tensorflow/blob/d8ce9f9c301d021a69953134185ab728c1c248d3/tensorflow/core/kernels/reshape_op.cc#L49-L70

**Comment by 372046933** [comment]
>TL;DR: There is no kernel for int32-valued Variable ops on GPU, but other element types (float16/float32/float64) are supported on GPU.

https://stackoverflow.com/a/37452938/1319804

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
